### DISCUSSION TITLE
Using Junctions on Windows

### DISCUSSION THREAD

**Discussion by quintesse** [jbangdev/jbang#1805] [General]
**Background**:

JBang currently uses symbolic links for things like setting the default JDK to use and for it's `edit` command.

**Problem**

But on WIndows creating symbolic links needs elevated privileges, you either need to be an administrator or you need to enable [Developer Mode](https://learn.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development). Running JBang as an admin is not a viable option and not everybody is able or willing to enable Developer Mode just to run JBang.

So in Windows there exists an additional feature, called Junctions, which is similar to symbolic links and which doesn't need any elevated privileges, but it can only be used for directories, not files.

Using junctions would work just fine for our JDK management because that only uses symbolic links to folders but JBang's `edit` command uses symbolic links to files, so junctions won't work in that case.

**Solutions?**

* For a second I thought about getting rid of the symlink all together for JDK management and just have some kind of configuration value that selects the JDK to use, but then we'd lose the capability to easily switch between JDKs. You'd need shell restarts all the time, etc, which sucks (It also doesn't fix the `edit` issue). So not an option 
* One option would be to keep both junctions and symbolic links on Windows. Meaning that we could use junctions with non-admin/non-dev mode for the JDK management necessary for a basic JBang install while needing elevated rights only when using the `edit` command. This is obviously not ideal.
* Another option is to perhaps see if we can remove the requirement for symbolic links to files from the `edit` command, but I think it would hard or impossible to offer the same features.
* Any other ideas? ...

**Comment by jcrben** [comment]
Have you considered hardlinks on Windows for the edit? They break a lot more easily than a symlink but don't require admin privileges.

Also, there's this concept of executable shims https://docs.chocolatey.org/en-us/features/shim - not sure it would apply with a file.

**Reply by quintesse** [reply]
I didn't know you don't need admin rights for hard links, but the problem with them is that they need to be on the same drive. For the `edit` command we set up a temporary project in the user's temp folder, this would cause problems when the original files are on a different drive. We'd be forced to create temp folders on each drive or something like that.

And the linked files are source files, so shims won't work (still, thanks for the link, the information is very interesting!)

**Reply by jcrben** [reply]
fyi @quintesse I discovered another trick to create "symlinks" on Windows with msys2 https://www.msys2.org/docs/symlinks/

`MSYS=winsymlinks:sys` creates "fake" symlinks which work pretty well as long as I'm still inside the msys2 environment:

"If set to winsymlinks:sys, Cygwin creates symlinks as plain files with the system attribute, containing a magic cookie followed by the path to which the link points"

https://cygwin.com/cygwin-ug-net/using-cygwinenv.html 
![image](https://github.com/user-attachments/assets/f79aea69-7a7a-44c8-8bb2-ab346c3d17ce)

However this still doesn't work with `jbang jdk default <version` - I guess because it isn't using `ln -s` under the hood.
